Add ConsumerNotifier and ACL schema for PINs / P4Runtime

bocon13 opened the pull request:
* Add Bazel build support so that swss-common can be used
  in Bazel projects; this doesn’t change/impact the
  make-based build system
* Add ConsumerNotifier for response path
* Introduce interfaces for DBConnector, NotificationConsumer,
  ConsumerNotifier, NotificationProducer, ProducerStateTable
  to make it easier to test these classes
* Add AppDB ACL schema values between the P4RT app and P4Orch
* Add lots of new tests for notification classes

Submission containing materials of a third party:
    Copyright Google LLC; Licensed under Apache 2.0 

Co-authored-by: Akarsh Gupta <akarshgupta@google.com>
Co-authored-by: Ashish Singh <ashishksingh@google.com>
Co-authored-by: Jay Hu <jiarenhu@google.com>
Co-authored-by: Manali Kumar <manalikumar@google.com>
Co-authored-by: Richard Yu <richardyu@google.com>
Co-authored-by: Robert J. Halstead <rhalstea@google.com>
Co-authored-by: Runming Wu <runmingwu@google.com>
Co-authored-by: Srikishen Pondicherry Shanmugam <kishanps@google.com>
Co-authored-by: Vivek Ramamoorthy <vivekmoorthy@google.com>
Co-authored-by: Yilan Ji <yilanji@google.com>

Signed-off-by: Don Newton don@opennetworking.org

bocon13 commented:
cc: @mint570

bocon13 commented on tests/test_redis_ut.py:
```diff
@@ -561,3 +561,17 @@ def test_DBConnectFailure():
     with pytest.raises(RuntimeError):
         db = swsscommon.DBConnector(0, nonexisting_host, 0)
 
+
+def test_StateHelper():
+    sh = swsscommon.StateHelperManager_SystemSingleton()
+    ch = swsscommon.StateHelperManager_ComponentSingleton(
+        swsscommon.SystemComponent_kP4rt)
+    assert sh.GetSystemState() == swsscommon.SystemState_kInitializing
+    assert sh.IsSystemCritical() == False
+    assert ch.StateInfo().state == swsscommon.ComponentState_kInitializing
+    assert ch.ReportComponentState(
+        swsscommon.ComponentState_kError, "Reason") == True
+    time.sleep(1)
+    assert sh.GetSystemState() == swsscommon.SystemState_kCritical
+    assert sh.IsSystemCritical() == True
+    assert ch.StateInfo().state == swsscommon.ComponentState_kError
```
```suggestion
```

bocon13 commented on tests/saiaclschema_ut.cpp:
```diff
@@ -0,0 +1,92 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/mocks/mock_producer_state_table.h:
```diff
@@ -0,0 +1,40 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/mocks/mock_db_connector.h:
```diff
@@ -0,0 +1,46 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/mocks/mock_consumer_notifier.h:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/mocks/BUILD:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/fake_sonic_db_table_test.cc:
```diff
@@ -0,0 +1,288 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/fake_sonic_db_table.h:
```diff
@@ -0,0 +1,95 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/fakes/fake_sonic_db_table.cc:
```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/fake_producer_state_table.h:
```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/fakes/fake_producer_state_table.cc:
```diff
@@ -0,0 +1,56 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/fake_db_connector.h:
```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/fakes/fake_db_connector.cc:
```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/fake_consumer_notifier.h:
```diff
@@ -0,0 +1,49 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on tests/fakes/fake_consumer_notifier.cc:
```diff
@@ -0,0 +1,34 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on tests/fakes/BUILD:
```diff
@@ -0,0 +1,83 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
```suggestion
```

bocon13 commented on tests/Makefile.am:
```diff
@@ -34,6 +34,7 @@ tests_SOURCES = redis_ut.cpp                \
                 redisutility_ut.cpp         \
                 boolean_ut.cpp              \
                 status_code_util_test.cpp   \
+		saiaclschema_ut.cpp         \
```
```suggestion
                saiaclschema_ut.cpp         \
```

bocon13 commented on common/saiaclschema.h:
```diff
@@ -0,0 +1,94 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/saiaclschema.cpp:
```diff
@@ -0,0 +1,453 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/producerstatetableinterface.h:
```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/dbconnectorinterface.h:
```diff
@@ -0,0 +1,46 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/consumernotifierinterface.h:
```diff
@@ -0,0 +1,36 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/consumernotifier.h:
```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
```suggestion
```

bocon13 commented on common/consumernotifier.cpp:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
```

bocon13 commented on BUILD:
```diff
@@ -0,0 +1,44 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
```suggestion
```

bocon13 commented on common/producerstatetable.h:
```diff
@@ -37,15 +39,15 @@ class ProducerStateTable : public TableBase, public TableName_KeySet
-    // The batched methods are not virtual. They don't include (or use) op and
-    // prefix. They are written for specific use case only.
-    // The consumer logic (or batch size) might need to change if the producer
-    // does batched operations.
+    // The batched methods don't include (or use) op and prefix. They are
+    // written for specific use case only. The consumer logic (or batch size)
+    // might need to change if the producer does batched operations.
```
Review

mint570 reviewed (commented):
LGTM

bocon13 commented:
@qiluo-msft please take a look

bocon13 commented:
@donNewtonAlpha Move this to upstream

rhalstea commented on tests/fakes/fake_db_connector.cc:
```diff
@@ -0,0 +1,192 @@
+
+}  // namespace
+
+void FakeDBConnector::AddAppDbTable(const std::string& table_name,
```
This should be named AddSonicDbTable.

Originally we only fakes AppDb tables, later we used the same paradigm to fake others.

rhalstea reviewed (commented):
A lot of the fakes don't seem to be up to date with the more recent pins changes?

bocon13 committed:
Add ConsumerNotifier for PINS response path

* Add ConsumerNotifier for response path
* Introduce interfaces for DBConnector, NotificationConsumer,
  ConsumerNotifier, NotificationProducer, ProducerStateTable
  to make it easier to test these classes

Submission containing materials of a third party:
    Copyright Google LLC; Licensed under Apache 2.0

Co-authored-by: Akarsh Gupta <akarshgupta@google.com>
Co-authored-by: Ashish Singh <ashishksingh@google.com>
Co-authored-by: Jay Hu <jiarenhu@google.com>
Co-authored-by: Manali Kumar <manalikumar@google.com>
Co-authored-by: Richard Yu <richardyu@google.com>
Co-authored-by: Robert J. Halstead <rhalstea@google.com>
Co-authored-by: Runming Wu <runmingwu@google.com>
Co-authored-by: Srikishen Pondicherry Shanmugam <kishanps@google.com>
Co-authored-by: Vivek Ramamoorthy <vivekmoorthy@google.com>
Co-authored-by: Yilan Ji <yilanji@google.com>

Signed-off-by: Don Newton don@opennetworking.org

Commit:
Add new tests for notification classes

bocon13 committed:
Rename files

donNewtonAlpha committed:
Merge remote-tracking branch 'origin/pr5-response-path-tests' into pr3-acl-schema

bocon13 committed:
Add libgmock to pipelines test
